composable-tasks/src/compose_tasks.rs

```diff
@@ -0,0 +1,137 @@
+use core::convert::TryFrom;
+
+use crate::external_sc_interactions;
+
+multiversx_sc::imports!();
+multiversx_sc::derive_imports!();
+
+pub type PaymentsVec<M> = ManagedVec<M, EsdtTokenPayment<M>>;
+
+#[derive(TypeAbi, TopEncode, TopDecode, PartialEq, ManagedVecItem)]
+pub enum TaskType {
+    WrapEGLD,
+    UnwrapEGLD,
+    Swap,
+    RouterSwap,
+    SendEgldOrEsdt,
+}
+
+#[multiversx_sc::module]
+pub trait TaskCall:
+    external_sc_interactions::pair_actions::PairActionsModule
+    + external_sc_interactions::router_actions::RouterActionsModule
+    + external_sc_interactions::wegld_swap::WegldWrapModule
+{
+    #[payable("*")]
+    #[endpoint(composeTasks)]
+    fn compose_tasks(
+        &self,
+        min_expected_token_out: EgldOrEsdtTokenPayment,
+        tasks: MultiValueEncoded<MultiValue2<TaskType, ManagedVec<ManagedBuffer>>>,
+    ) {
+        let mut payment_for_next_task = self.call_value().egld_or_single_esdt();
+        let mut payments_to_return = PaymentsVec::new();
+
+        let caller = self.blockchain().get_caller();
+
+        for task in tasks.into_iter() {
+            let (task_type, args) = task.into_tuple();
+
+            let payment_for_current_task = payment_for_next_task.clone();
+
+            payment_for_next_task = match task_type {
+                TaskType::WrapEGLD => self.wrap_egld(payment_for_current_task),
+                TaskType::UnwrapEGLD => self.unwrap_egld(payment_for_current_task),
+                TaskType::Swap => {
```

Comment:
Maybe we could refactor this to have a function for each step.
```
TaskType::WrapEGLD => self.wrap_egld(args),
TaskType::UnwrapEGLD => self.unwrap_egld(args),
TaskType::Swap => self.swap(args),
TaskType::RouterSwap => self.router_swap(args),
```

This way, it will be easier to read the code. It will be a bit tricky to see how to pass the arguments.